dashboard/app/controllers/admin_users_controller.rb

```diff
@@ -101,33 +101,59 @@ def manual_pass
     redirect_to :manual_pass_form
   end
 
+  # display the user search form, users found, and permissions for first user found
+  # grant / revoke permissions actions redirect back to this form to display the updated permission list
   def permissions_form
+    @users = Array.new # list of users that match search
+    @permissions = Array.new
+    search_term = user_permission_params[:search_term]
+    if search_term =~ /^\d+$/
+      user_id = search_term
+      @users = User.where(id: user_id)
+    elsif search_term
+      email = search_term
+      hashed_email = User.hash_email(email)
+      @users = User.where(hashed_email: hashed_email)
+    end
+    @user = @users.first if @users
+    @permissions = @user.permissions if @user
   end
 
-  # Grants the indicated permission to the indicated user. Expects params[:email] and
-  # params[:permission] to be populated.
   def grant_permission
-    permission = params[:permission]
-    user = User.find_by_email_or_hashed_email params[:email]
+    user_id = user_permission_params[:user_id]
+    permission = user_permission_params[:user_permission_id]
+    @user = User.find(user_id)
 
-    unless user && user.teacher?
-      flash[:alert] = "FAILED: user #{params[:email]} could not be found or was not a teacher"
-      redirect_to :permissions_form
+    unless @user && @user.teacher?
+      flash[:alert] = "FAILED: user #{user_permission_params[:user_id]} could not be found or was not a teacher"
+      redirect_to action: "permissions_form", search_term: user_id
       return
     end
 
     if permission == 'admin'
-      if user.sections_as_student.count > 0
-        flash[:alert] = "FAILED: user #{user.email} NOT granted as user has sections_as_students"
+      if @user.sections_as_student.count > 0
```

Comment:
I agree with Asher. I realize this check was here before this PR, but since we're working here, let's remove it in favor of the model validations. We don't need strict validations here.

The bang methods, e.g. `@user.update!` raise an `ActiveRecord::RecordInvalid` when the record is invalid. Alternatively the non-bang `update` returns bool and won't save if it's invalid, so you could either do

```ruby
begin
  @user.update!(admin: true)
rescue ActiveRecord::RecordInvalid
  flash...
end
```
or
```ruby
unless @user.update(admin: true)
  flash...
end
```

Note the code as it is will raise an unhandled error and blow up (from the model validation) if we try to assign admin to a student.